### PR TITLE
add focus & resize window functions

### PR DESCRIPTION
**Pull request by ekrenzin**
# Title: PR: Add `focusWindow` and `resizeWindow` Functionality

## Description

This pull request introduces new functionality to the window manager. With the modifications made in this PR, users can now focus on a specific window and resize it using its window handle. The implementation covers all supported platforms - Windows, macOS, and Linux. 

## Changes

1. **index.d.ts**: Declared new exported functions `focusWindow` and `resizeWindow`.
2. **permissionCheck.js**: Added `focusWindow` and `resizeWindow` to permission checks.
3. **src/linux/window_manager.cc**: Implemented the functionality of `focusWindow` and `resizeWindow` for Linux using Xlib.
4. **src/macos/window_manager.mm**: Implemented the functionality of `focusWindow` and `resizeWindow` for macOS using Objective-C++ and the Cocoa API.
5. **src/main.cc**: Added the wrapper functions for `focusWindow` and `resizeWindow`, enabling them to be called from JavaScript.
6. **src/win32/window_manager.cc**: Implemented the functionality of `focusWindow` and `resizeWindow` for Windows using Win32 API.
7. **src/window_manager.h**: Added the function declarations for `focusWindow` and `resizeWindow`.

### New Functions

- `focusWindow(handle: number): void;`
Focuses the window specified by the provided window handle. Brings the window to the foreground and gives it input focus.

- `resizeWindow(handle: number, width: number, height: number): void;`
Resizes the window specified by the provided window handle to the given width and height.

## Testing

Please test these changes by trying to focus and resize the window using their respective new functions on all supported platforms. Verify that the functions return appropriate boolean values indicating the success of the operation. 

## Impact

These changes should not impact any existing functionality in the application. Instead, they provide the users with new capabilities of focusing and resizing windows from their application.

## Open Issue

https://github.com/nut-tree/libnut-core/issues/18

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -51,5 +51,7 @@ export function getWindows(): number[];
 export function getActiveWindow(): number;
 export function getWindowRect(handle: number): Rect;
 export function getWindowTitle(handle: number): string;
+export function focusWindow(handle: number): void;
+export function resizeWindow(handle: number, width: number, height: number): void;
```
Comment (s1hofmann) [inline index.d.ts:55]: **Suggestion**: The size parameter could be an object of shape

```js
{
  x: number;
  y: number;
  width: number;
  height: number
}
```

This way we could perform resizing + repositioning (which is a common combination IMHO).

WDYT?

File: src/window_manager.h [hunk 2]
```diff
@@ -31,6 +31,25 @@ std::string getWindowTitle(const WindowHandle windowHandle);
...
+
+/**
+ * `resizeWindow` resizes the window specified by its window handle to the given width and height.
+ * The respective window handle may be acquired via `getWindows` or `getActiveWindow`.
+ * @param windowHandle The window handle of the window to be resized.
+ * @param width The new width of the window.
+ * @param height The new height of the window.
+ * @return Returns a boolean indicating whether the window resize operation was successful.
+ */
+bool resizeWindow(const WindowHandle windowHandle, int width, int height);
```
Comment (s1hofmann) [inline src/window_manager.h:51]: **Praise**: Doc strings, awesome! 💯

File: index.d.ts (hunk 1 above)
Comment (ekrenzin) [reply index.d.ts:55]: Yeah great point, I completely agree!

Comment (ekrenzin) [reply index.d.ts:55]: I'm going to use the Rect interface from line 25

File: src/main.cc [hunk 3]
```diff
@@ -644,6 +644,53 @@ Napi::String _getWindowTitle(const Napi::CallbackInfo &info) {
...
+    int64_t height = 0;
+
+    if (rectObj.Has("x"))
+        x = rectObj.Get("x").As<Napi::Number>().Int64Value();
+    if (rectObj.Has("y"))
+        y = rectObj.Get("y").As<Napi::Number>().Int64Value();
+    if (rectObj.Has("width"))
+        width = rectObj.Get("width").As<Napi::Number>().Int64Value();
+    if (rectObj.Has("height"))
+        height = rectObj.Get("height").As<Napi::Number>().Int64Value();
```
Comment (s1hofmann) [inline src/main.cc:681]: **Suggestion**: I'd prefer to throw an error if we don't receive a proper rect object. Resizing a window to (0,0,0,0) makes no sense IMHO

File: src/main.cc [hunk 4]
```diff
@@ -644,6 +644,53 @@ Napi::String _getWindowTitle(const Napi::CallbackInfo &info) {
...
+
+    windowRect.origin.x = x;
+    windowRect.origin.y = y;
+    windowRect.size.width = width;
+    windowRect.size.height = height;
+
+    resizeWindow(windowHandle, windowRect);
+
+    return Napi::Boolean::New(env, true);
```
Comment (s1hofmann) [inline src/main.cc:690]: **Suggestion**: `resizeWindow` returns a boolean, so we should return this result here.

**Review by s1hofmann — CHANGES_REQUESTED** [review]

**Review by s1hofmann — APPROVED** [review]

**Pull request merged**